src/transaction/transaction.py

```diff
@@ -0,0 +1,255 @@
+                - unspent transaction object's public key script (hashed pubkey)
+                - this transaction's public key script (s)?
+                - this transaction's output amount (s)?
+                and check to see if this transaction's signature can be 
+                verified by the now authorized signature script key.             
+
+        :return: True if the transaction is valid, false otherwise
+        """
+        authentic = False
+        for tnx_input in self.inputs:  # for each referenced input
+            utxo = find_unspent_output(tnx_input['transaction_id'],
+                                       tnx_input['output_index'],
+                                       tnx_input['block_hash'])
+
+            sig_key = SHA256.new(  # get this transaction's unlock public key
+                tnx_input['unlock']['public_key'].encode()
+            ).hexdigest()
+
+            if sig_key == utxo['address']:  # if this node is the recipient of
+                # the previous utxo
+                transaction_message = SHA256.new((  # transaction message
+                    str(tnx_input['transaction_id']) +  # input id
+                    str(tnx_input['output_index']) +  # output index
+                    str(utxo['address']) +  # hashed public key as address
+                    str(self.outputs)
+                ).encode('utf-8'))
+
+                ecc_key = import_public_key(tnx_input['unlock']['public_key'])
+                signature = tnx_input['unlock']['signature']
+                decoded = base64.b64decode(signature.encode())
+                verifier = DSS.new(ecc_key, 'fips-186-3')
+                try:
+                    verifier.verify(transaction_message, decoded)
+                    authentic = True
+                except ValueError:
+                    authentic = False
+        # TODO: raise error if false
+        return authentic
+
+    def get_transaction_id(self):
+        """
+        Calculate the transaction id. 
+        :return: the transaction id of this transaction
+        """
+        transaction = {
+            "input_count": self.input_count,
+            "inputs": self.inputs,
+            "output_count": self.output_count,
+            "outputs": self.outputs
+        }
+        txid = SHA256.new(
+            str(transaction).encode()
+        ).hexdigest()
+        self.transaction_id = txid
+        return txid
+
+    def get_data(self):
+        """
+        Get a dict representation of a transaction with the
+        transaction id as the key
+        :return: a dict representation of the transaction object.
+        This would typically be the value in a tx key-value
+        """
+        transaction = {
+            "input_count": self.input_count,
+            "inputs": self.inputs,
+            "output_count": self.output_count,
+            "outputs": self.outputs
+        }
+        return transaction
+
+
+def create_coinbase_tx(reward_amount):
+    """
```

Comment:
Will every transaction have one of these? If so we can get rid of this function and just have it be automatically added right before the transaction is signed.

Reply:
Not every transaction will have this, but every block will. This function essentially returns a Transaction where there are no inputs and only one output going to the miner's address